GH-2412: Improved JoinClassifier for unbound values

Aklakan opened the pull request:
GitHub issue resolved #2412

Pull request Description: Additional heuristics and test cases in JoinClassifier and VarFinder for dealing with unbound values.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:
```diff
@@ -351,9 +389,26 @@ public void visit(OpProject opProject) {
+                }
+            }
+            defines.addAll(definedVars);
+            optDefines.addAll(optVars);
```
Are the arguments against treating unbound table values as optDefines?

rvesse replied:
No, makes sense to me

Aklakan replied:
Added variable to optDefines if it is not considered defined.

Aklakan commented on jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java:
```diff
@@ -204,7 +204,17 @@ public void dynamic_dft_7() {
     }
-    
+
+    @Test
```
This class is not an ideal fit for the join tests. Any suggestion?

rvesse replied:
Can you not make the `testCount()` method `public static` and then create a new test class that simply calls it?

Aklakan replied:
With my question I was mainly referring to whether I have overlooked an already existing suitable place for these tests - if there is none so far I could of course create a new class and register it with the test suite.

Aklakan replied:
There is now a AbstractTestQueryExec hierarchy analogous to AbstractTestDynamicDataset.

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
+                if ( e != null ) {
+                    boolean isLikelyToBeDefined = isLikelyToBeDefined(e);
+                    if (isLikelyToBeDefined) {
+                        defines.add(v) ; // Expression may eval to error -> unset?
```
If the variable may be unbound it probably still needs to be tracked - e.g. in optDefines?

afs replied:
Yes - that seems likely. 
Not putting it in is quite safe.
Doing so and tests all pass.

Aklakan replied:
Done

Aklakan commented:
I just removed some left-over code that had slipped into this PR.
The proposed changes are ready for discussion.

rvesse commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:
```diff
@@ -351,9 +389,26 @@ public void visit(OpProject opProject) {
+                Iterator<Var> varIt = definedVars.iterator();
+                while (varIt.hasNext()) {
+                    Var var = varIt.next();
+                    if (!row.contains(var)) {
```
This seems overly aggressive at first glance.  If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked. What happens if you have multiple rows with some values and some `UNDEF`, is that going to create bad results if we allow linearisation?

Perhaps a different solution is needed around handling of linear joins for `OpTable` to just never substitute over an `UNDEF` though guess that might be significantly more involved?

Aklakan replied:
> If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked.

Its per variable - so if there is an UNDEF value for a variable then linearization for that variable gets blocked.
Conversely, if the variable is defined in all rows its added to the defined set.
I was creating test cases for skewed multi-variable joins (i.e. joins on multiple variables where some of them are unbound) when I noticed that I get incorrect results.

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
+                ExprFunction fn = expr.getFunction();
+                List<Expr> args = fn.getArgs();
+                if (fn instanceof E_Coalesce) {
+                    result = args.stream().anyMatch(this::isLikelyToBeDefined);
```
Done

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
+                if (fn instanceof E_Coalesce) {
+                    result = args.stream().anyMatch(this::isLikelyToBeDefined);
+                } else {
+                    result = args.stream().allMatch(this::isLikelyToBeDefined);
```
Done

Aklakan commented:
I think I addressed all changed. You probably still want to check the AbstractTestQueryExec / TestQueryExec* classes that I added.

afs commented:
Looks good.

Attached are some more tests `TestVarFinder2` for `VarFinder` directly.
If you agree with my understanding, please add the file and add to `TS_Algebra`.
(5 tests fail in Jena without this PR).
[TestVarFinder2.java.txt](https://github.com/apache/jena/files/15049216/TestVarFinder2.java.txt)

One clearing thing:
`AbstractTestDynamicDataset` - the `join_*` test are still there.

Aklakan commented:
The test cases in TestVarFinder2 look good to me.
I added one more assertion to `varfind_extend_6`: `assertSetVars(vf.getOpt(), var_x);`

> AbstractTestDynamicDataset - the join_* test are still there.

Whoops, removed.

afs commented:
Thanks!
